packages/firebase_storage/lib/src/upload_task.dart

```diff
@@ -1,81 +1,185 @@
+            new StorageTaskEvent._(m['type'], m['snapshot']))
+        .map<StorageTaskEvent>((StorageTaskEvent e) {
+          _resetState();
+          switch (e.type) {
+            case StorageTaskEventType.progress:
+              isInProgress = true;
+              break;
+            case StorageTaskEventType.resume:
+              isInProgress = true;
+              break;
+            case StorageTaskEventType.pause:
+              isPaused = true;
+              break;
+            case StorageTaskEventType.success:
+              isSuccessful = true;
+              isComplete = true;
+              _completer.complete(e.snapshot);
+              _downloadUrl.complete(e.snapshot.downloadUrl);
+              break;
+            case StorageTaskEventType.failure:
+              isComplete = true;
+              if (e.snapshot.error == StorageError.canceled) {
+                isCanceled = true;
+              }
+              _completer.complete(e.snapshot);
+              _downloadUrl.complete(e.snapshot.downloadUrl);
+              break;
+          }
+          lastSnapshot = e.snapshot;
+          _controller.add(e);
+          return e;
+        })
+        .firstWhere((StorageTaskEvent e) =>
+            e.type == StorageTaskEventType.success ||
+            e.type == StorageTaskEventType.failure)
+        .then<StorageTaskSnapshot>((StorageTaskEvent event) => event.snapshot);
+  }
 
-  StorageUploadTask._(this._firebaseStorage, this._path, this._metadata);
-  Future<void> _start();
+  void _resetState() {
+    isCanceled = false;
+    isComplete = false;
+    isInProgress = false;
+    isPaused = false;
+    isSuccessful = false;
+  }
+
+  /// Pause the upload
+  void pause() => FirebaseStorage.channel.invokeMethod(
+        'UploadTask#pause',
+        <String, dynamic>{
+          'app': _storage.app?.name,
+          'bucket': _storage.storageBucket,
+          'handle': _handle,
+        },
+      );
+
+  /// Resume the upload
+  void resume() => FirebaseStorage.channel.invokeMethod(
+        'UploadTask#resume',
+        <String, dynamic>{
+          'app': _storage.app?.name,
+          'bucket': _storage.storageBucket,
+          'handle': _handle,
+        },
+      );
 
-  Completer<UploadTaskSnapshot> _completer =
-      new Completer<UploadTaskSnapshot>();
-  Future<UploadTaskSnapshot> get future => _completer.future;
+  /// Cancel the upload
+  void cancel() => FirebaseStorage.channel.invokeMethod(
+        'UploadTask#cancel',
+        <String, dynamic>{
+          'app': _storage.app?.name,
+          'bucket': _storage.storageBucket,
+          'handle': _handle,
+        },
+      );
 }
 
-class StorageFileUploadTask extends StorageUploadTask {
-  final File _file;
-  StorageFileUploadTask._(this._file, FirebaseStorage firebaseStorage,
-      String path, StorageMetadata metadata)
-      : super._(firebaseStorage, path, metadata);
+class _StorageFileUploadTask extends StorageUploadTask {
```

Comment:
Are we sure we want to do this? This will introduce some ambiguity for the end user.  Why expose `StorageFileUploadTask` and `StorageDataUploadTask` when the only functionality that would be needed is found in the base class, `StorageUploadTask`?